File: Contour-Detection-using-OpenCV/cpp/channel_experiments/channel_experiments.cpp
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include <array>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#ifndef TUTORIAL_INPUT_DIR
#define TUTORIAL_INPUT_DIR "../../input"
#endif

namespace fs = std::filesystem;

struct Options {
    fs::path input = fs::path(TUTORIAL_INPUT_DIR) / "image_1.jpg";
    fs::path output_dir = fs::current_path();
    bool display = true;
    bool validate = false;
};

Options parse_options(int argc, char** argv) {
    Options options;
    for (int i = 1; i < argc; ++i) {
        const std::string argument = argv[i];
        if ((argument == "--input" || argument == "--output-dir") && i + 1 >= argc) {
            throw std::invalid_argument(argument + " requires a path");
        }
        if (argument == "--input") {
            options.input = argv[++i];
        } else if (argument == "--output-dir") {
            options.output_dir = argv[++i];
        } else if (argument == "--no-display") {
            options.display = false;
        } else if (argument == "--validate") {
            options.validate = true;
        } else {
            throw std::invalid_argument("Unknown argument: " + argument);
        }
    }
    return options;
}

void show_image(const std::string& title, const cv::Mat& image, bool display) {
    if (display) {
        cv::imshow(title, image);
        cv::waitKey(0);
        cv::destroyAllWindows();
    }
}

int main(int argc, char** argv) {
    try {
        const Options options = parse_options(argc, argv);
        const cv::Mat image = cv::imread(options.input.string());
        if (image.empty()) {
            throw std::runtime_error("Could not read input image: " + options.input.string());
        }
        fs::create_directories(options.output_dir);

        std::vector<cv::Mat> channels;
        cv::split(image, channels);
        const std::array<std::string, 3> names{"blue", "green", "red"};
        std::array<std::size_t, 3> counts{};

        for (std::size_t index = 0; index < names.size(); ++index) {
            std::vector<std::vector<cv::Point>> contours;
            std::vector<cv::Vec4i> hierarchy;
            cv::findContours(
                channels[index].clone(), contours, hierarchy, cv::RETR_TREE,
                cv::CHAIN_APPROX_NONE
            );

            cv::Mat rendered = image.clone();
            cv::drawContours(
                rendered, contours, -1, cv::Scalar(0, 255, 0), 2, cv::LINE_AA
            );
            show_image("Contours from the " + names[index] + " channel", rendered,
                       options.display);

            const fs::path output = options.output_dir / (names[index] + "_channel.jpg");
            if (!cv::imwrite(output.string(), rendered)) {
                throw std::runtime_error("Could not write output image: " + output.string());
            }
            counts[index] = contours.size();
            std::cout << names[index] << ": contours=" << contours.size() << '\n';
        }

        if (options.validate) {
            if (!(counts[0] > 0 && counts[0] < counts[1] && counts[1] < counts[2])) {
                throw std::runtime_error(
                    "Expected the bundled image to produce 0 < blue < green < red contours"
                );
            }
            std::cout << "Validation passed\n";
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}

File: Contour-Detection-using-OpenCV/python/channel_experiments/channel_experiments.py
"""Show why raw B, G, and R channels are poor contour inputs."""

from __future__ import annotations

import argparse
from pathlib import Path

import cv2


DEFAULT_INPUT = Path(__file__).resolve().parents[2] / "input" / "image_1.jpg"


def show_image(title: str, image, display: bool) -> None:
    if display:
        cv2.imshow(title, image)
        cv2.waitKey(0)
        cv2.destroyAllWindows()


def run(input_path: Path, output_dir: Path, display: bool = True) -> dict[str, int]:
    image = cv2.imread(str(input_path))
    if image is None:
        raise FileNotFoundError(f"Could not read input image: {input_path}")

    output_dir.mkdir(parents=True, exist_ok=True)
    counts: dict[str, int] = {}

    # OpenCV stores color images in B, G, R order.
    for name, channel in zip(("blue", "green", "red"), cv2.split(image)):
        contours, _ = cv2.findContours(
            channel.copy(), cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE
        )
        rendered = image.copy()
        cv2.drawContours(rendered, contours, -1, (0, 255, 0), 2, cv2.LINE_AA)

        show_image(f"Contours from the {name} channel", rendered, display)
        output_path = output_dir / f"{name}_channel.jpg"
        if not cv2.imwrite(str(output_path), rendered):
            raise OSError(f"Could not write output image: {output_path}")

        counts[name] = len(contours)
        print(f"{name}: contours={len(contours)}")

    return counts


def validate(counts: dict[str, int]) -> None:
    if not all(count > 0 for count in counts.values()):
        raise RuntimeError("Every color channel should produce at least one contour")
    if not counts["blue"] < counts["green"] < counts["red"]:
        raise RuntimeError(
            "Expected the bundled image to produce blue < green < red contour counts"
        )


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", type=Path, default=DEFAULT_INPUT)
    parser.add_argument("--output-dir", type=Path, default=Path.cwd())
    parser.add_argument("--no-display", action="store_true", help="Skip GUI windows")
    parser.add_argument(
        "--validate", action="store_true", help="Validate bundled-image invariants"
    )
    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()
    result = run(args.input, args.output_dir, display=not args.no_display)
    if args.validate:
        validate(result)
        print("Validation passed")
